packages/lodestar/src/chain/chain.ts

```diff
@@ -84,6 +85,7 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
     this.attestationProcessingQueue = queue(async (task: Function) => {
       await task();
     }, 1);
+    this.attesationWaitingQueue = new FastPriorityQueue((a: Attestation, b: Attestation) => a.data.slot < b.data.slot);
```

Comment:
Why do you need them ordered by slot and how is that different from processing queue?

I think we just need to add check to prevent processing attestation if we don't yet have targeted block

Reply:
At each slot validators will post attestation for that slot so it's reasonable to process attestation slot by slot.

The processing queue will process attestation without to check for anything. The waiting queue sits before the processing queue and we stop in the middle if any attestation don't meet our checks below, just like the block queue.

We do need this waiting queue because the spec said "Attestations target be for a known block. If target block is unknown, delay consideration until the block is found"

Reply:
YOu are assuming we are dealing with local nodes and validators. In reality we might receive attestation but not a block or we receive attestation too late. There can also be multiple attestations at same slot for different block etc.

According to spec we have two scenarios:
1. We received attestation for known block -> we include it in fork choice
2. We received attestation for unknown block -> emit unknownBlockRoot event -> include block in forkchoice -> check in pending attestation if some are targeting that block -> include attestation

Reply:
So I suppose the approach is:
+ Don't need a new attestation queue
+ Create `Record<blockRoot, Attestation> pendingAttestations` inside chain.ts (or any similar ways)
+ Add check to prevent processing attestations, if it does not pass the check then put it in `pendingAttestations` above
+ When we receive block, get all pending attestations respective to that block and add to processing queue. Remove them from `pendingAttestations`

Is that correct?

Reply:
Yeah, few advice's:
- try to extract those attestation processing outside of chain module (it will be easier to test and maintain)
- If you implement that, I don't think we need attestation processing queue because we have that queue just so we don't process attestation before block which should't happen now